sdk/core/Azure.Core/src/Shared/NoValueResponseOfT.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure
+{
+#pragma warning disable SA1649 // File name should match first type name
+    internal sealed class NoValueResponse<T> : NullableResponse<T>
+#pragma warning restore SA1649 // File name should match first type name
+    {
+        private readonly Response _response;
+
+        public NoValueResponse(Response response)
+        {
+            if (response == null)
+            {
+                throw new ArgumentNullException(nameof(response));
+            }
+            _response = response;
+        }
+
+        /// <inheritdoc />
+        public override bool HasValue => false;
+
+        public override T Value
+        {
+            get
+            {
+                throw new ResponseBodyNotFoundException(_response.Status, _response.ReasonPhrase);
+            }
+        }
+
+        public override Response GetRawResponse() => _response;
+
+        public override string ToString()
+        {
+            return $"Status: {GetRawResponse().Status}, Service returned no content";
+        }
+
+#pragma warning disable CA1064 // Exceptions should be public
+        private class ResponseBodyNotFoundException : Exception
```

Comment:
This name seems to correspond to the `NoBodyResponse` above ... it's possible that a `NullableResponse<T>` could have a body, but no value (i.e. if the payload failed to deserialize, or if the body contained an error).  It seems like another name might be better, or would it make sense to return an InvalidOperationException, since that's what `Nullable<T>` does?

Reply:
> it's possible that a NullableResponse<T> could have a body, but no value (i.e. if the payload failed to deserialize, or if the body contained an error).

I don't think this type is intended to represent those scenarios. We should throw in those examples.

>  It seems like another name might be better, or would it make sense to return an InvalidOperationException, since that's what Nullable<T> does?

I don't feel strongly about the name since it's private, but I like using a known exception type as you suggest.